### ISSUE TITLE
Only 2 gz files can be displayed when the latest version 32 is loaded

### ISSUE THREAD

**Issue by wujie1314** [niivue/niivue#501]
With BUG
Only 2 gz files can be displayed when the latest version 32 is loaded

**Comment by neurolabusc** [comment]
Unable to replicate. This script shows 4 voxel based images overlaid with NiiVue:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>NiiVue</title>
  </head>
  <body>
    <canvas id="gl" width="1024" height="640"></canvas>
  </body>
<script src="https://unpkg.com/@niivue/niivue@0.32.0/dist/niivue.umd.js"></script>
<script>
 var root = "https://niivue.github.io/niivue-demo-images/"
 var volumeList = [
  {url: root+"chris_t1.nii.gz", colorMap:"gray"},
  {url: root+"mni152.nii.gz", colorMap:"blue", opacity:0.2},
  {url: root+"spmMotor.nii.gz", colorMap:"red", opacity:0.5, cal_min: 1, cal_max: 5},
  {url: root+"spmMotor.nii.gz", colorMap:"green", opacity:0.5, cal_min: 4, cal_max: 5},
 ]
 var nv = new niivue.Niivue({isResizeCanvas: false})
 nv.attachTo("gl") 
 nv.loadVolumes(volumeList)
 nv.opts.isColorbar = true
</script>
```

![overlay](https://user-images.githubusercontent.com/8930807/216015865-90c94289-4493-408f-9d29-d8621e6cb81a.jpg)

Perhaps you can provide a script that illustrates your issue.

**Comment by neurolabusc** [comment]
In investigating this issue, I did notice that adding overlays can restrict the bounding box of previous overlays. While it is typical that all overlays share a common bounding box (and we expect that all voxel overlays are constrained by the bounding box of the background image), I have relaxed the constraint between overlays.

The script below illustrates the previous behavior, where the (misregistered) green overlay clips the signal of the red overlay.

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>NiiVue</title>
  </head>
  <body>
    <canvas id="gl" width="1024" height="640"></canvas>
  </body>
<script src="https://niivue.github.io/niivue/features/niivue.umd.js"></script>
<script>
 var root = "https://niivue.github.io/niivue-demo-images/"
 var volumeList = [
  {url: root+"spm152.nii.gz", colorMap:"gray"},
  {url: root+"spmMotor.nii.gz", colorMap:"red", opacity:0.5, cal_min: 1, cal_max: 5},
  {url: root+"CT_AVM.nii.gz", colorMap:"green", opacity:0.5, cal_min: 4, cal_max: 5},
 ]
 var nv = new niivue.Niivue({isResizeCanvas: false})
 nv.attachTo("gl") 
 nv.loadVolumes(volumeList)
 nv.opts.isColorbar = true
</script>
```

![relax2](https://user-images.githubusercontent.com/8930807/216022292-56c1f39d-93b7-463e-9c80-bd0ecd021769.jpg)

The latest commit updates the orientation shader to relax this constraint.

**Comment by hanayik** [comment]
@neurolabusc , can you open a new PR so we can get that commit merged into main?

**Comment by wujie1314** [comment]
[amos_0004_1.nii.gz](https://github.com/niivue/niivue/files/10576986/amos_0004_1.nii.gz)
[amos_0004_2.nii.gz](https://github.com/niivue/niivue/files/10576988/amos_0004_2.nii.gz)
[amos_0004_3.nii.gz](https://github.com/niivue/
[amos_0004_4.nii.gz](https://github.com/niivue/niivue/files/10576990/amos_0004_4.nii.gz)
niivue/files/10576989/amos_0004_3.nii.gz)

There is still a problem after trying the 32 version today. These files can be loaded normally after only 2 files are loaded together and the 30 version is returned

**Comment by neurolabusc** [comment]
Everything works fine here. Here is an example of what you get if you copy your images to the /niivue/src folder and run `npm run dev`:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,initial-scale=1.0">
  <title>NiiVue</title>
  <link rel="stylesheet" href="niivue.css">
</head>
<body>
    <noscript>
      <strong>niivue requires JavaScript.</strong>
    </noscript>
    <main>
      <canvas id="gl1"></canvas>
    </main>
    <script type="module" async>
      import { Niivue } from './niivue.js'
      import {NVImage} from './nvimage.js'
      import {NVMesh} from './nvmesh.js'
      var volumeList1 = [
        { url: "./amos_0004_1.nii.gz", colorMap: "Gray", },
        { url: "./amos_0004_2.nii.gz", colorMap: "Red",  },
        { url: "./amos_0004_3.nii.gz", colorMap: "Green",  },
        { url: "./amos_0004_4.nii.gz", colorMap: "Blue",  },
        ]
      var nv1 = new Niivue()
      nv1.setRadiologicalConvention(false)
      nv1.attachTo('gl1')
      nv1.setSliceType(nv1.sliceTypeMultiplanar)
      nv1.opts.dragMode = nv1.dragModes.pan
      await nv1.loadVolumes(volumeList1)
    nv1.opts.isColorbar = true
    nv1.setSliceType(nv1.sliceTypeRender)
   </script>
</body>
</html>
```
<img width="654" alt="Screen Shot 2023-02-03 at 11 35 09 PM" src="https://user-images.githubusercontent.com/8930807/216724184-2f113101-1f9b-4ffa-87bf-55c883b71c04.png">

I suspect your issue is that you are trying to load a URL where you have not set the CORS policy. You can see this by setting the path of one of your files to `https://github.com/niivue/niivue/files/10576986/amos_0004_1.nii.gz` and looking at the output of Chrome Developer Tools:

```
index.html:1 Access to script at 'file:///Users/chrisrorden/Desktop/niivue.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.
```

You can not read files from another github,com URL, rather you can read from a github.io URL.  For example, you can read the file `https://niivue.github.io/niivue-demo-images/chris_PD.nii.gz`. You can change a github repository to have a github.io by going to the Github repository's `Settings` and going to the `Pages` tab.

**Comment by wujie1314** [comment]
I used "@ nivue/nivue": "^ 0.32.0". If there is a problem, I will return to "@ nivue/nivue": "^ 0.30.0". There is no problem. The code has not been modified  
2023 2.6   Still, the problem can't be loaded

**Comment by wujie1314** [comment]
<script src="https://unpkg.com/@niivue/niivue@0.30.0/dist/niivue.umd.js"></script>
![image](https://user-images.githubusercontent.com/41436763/216871428-14ebc8da-c985-4d0d-a686-10908d2c58fb.png)
 <script src="https://unpkg.com/@niivue/niivue@0.32.0/dist/niivue.umd.js"></script>
![image](https://user-images.githubusercontent.com/41436763/216871591-f68bee06-1bba-465f-bdc3-1b0199b21e36.png)

code:
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>NiiVue</title>
  </head>
  <body>
    <canvas id="gl" width="640" height="640"></canvas>
  </body>

  <script src="https://unpkg.com/@niivue/niivue@0.32.0/dist/niivue.umd.js"></script>
<script type="module" async>
 var volumeList = [
     {
         url: './amos_0004_1.nii.gz',colorMap: "Green",
     },
     {
         url: './amos_0004_2.nii.gz',colorMap: "red",
     },
     {
         url: './amos_0004_3.nii.gz',colorMap: "Blue"
     },
 ]
 var nv = new niivue.Niivue({isResizeCanvas: false})
 nv.attachTo('gl')
 await nv.loadVolumes(volumeList)
 nv.opts.isColorbar = true
 nv.setSliceType(nv.sliceTypeRender)
</script>
</html>

**Comment by wujie1314** [comment]
> Everything works fine here. Here is an example of what you get if you copy your images to the /niivue/src folder and run `npm run dev`:
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
> <head>
>   <meta charset="utf-8">
>   <meta http-equiv="X-UA-Compatible" content="IE=edge">
>   <meta name="viewport" content="width=device-width,initial-scale=1.0">
>   <title>NiiVue</title>
>   <link rel="stylesheet" href="niivue.css">
> </head>
> <body>
>     <noscript>
>       <strong>niivue requires JavaScript.</strong>
>     </noscript>
>     <main>
>       <canvas id="gl1"></canvas>
>     </main>
>     <script type="module" async>
>       import { Niivue } from './niivue.js'
>       import {NVImage} from './nvimage.js'
>       import {NVMesh} from './nvmesh.js'
>       var volumeList1 = [
>         { url: "./amos_0004_1.nii.gz", colorMap: "Gray", },
>         { url: "./amos_0004_2.nii.gz", colorMap: "Red",  },
>         { url: "./amos_0004_3.nii.gz", colorMap: "Green",  },
>         { url: "./amos_0004_4.nii.gz", colorMap: "Blue",  },
>         ]
>       var nv1 = new Niivue()
>       nv1.setRadiologicalConvention(false)
>       nv1.attachTo('gl1')
>       nv1.setSliceType(nv1.sliceTypeMultiplanar)
>       nv1.opts.dragMode = nv1.dragModes.pan
>       await nv1.loadVolumes(volumeList1)
>     nv1.opts.isColorbar = true
>     nv1.setSliceType(nv1.sliceTypeRender)
>    </script>
> </body>
> </html>
> ```
> 
> <img alt="Screen Shot 2023-02-03 at 11 35 09 PM" width="654" src="https://user-images.githubusercontent.com/8930807/216724184-2f113101-1f9b-4ffa-87bf-55c883b71c04.png">
> 
> I suspect your issue is that you are trying to load a URL where you have not set the CORS policy. You can see this by setting the path of one of your files to `https://github.com/niivue/niivue/files/10576986/amos_0004_1.nii.gz` and looking at the output of Chrome Developer Tools:
> 
> ```
> index.html:1 Access to script at 'file:///Users/chrisrorden/Desktop/niivue.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.
> ```
> 
> You can not read files from another github,com URL, rather you can read from a github.io URL. For example, you can read the file `https://niivue.github.io/niivue-demo-images/chris_PD.nii.gz`. You can change a github repository to have a github.io by going to the Github repository's `Settings` and going to the `Pages` tab.

Do not import local js and use online ones

**Comment by wujie1314** [comment]
<script src="https://unpkg.com/@niivue/niivue@0.30.0/dist/niivue.umd.js"></script>
![image](https://user-images.githubusercontent.com/41436763/216871428-14ebc8da-c985-4d0d-a686-10908d2c58fb.png)
 <script src="https://unpkg.com/@niivue/niivue@0.32.0/dist/niivue.umd.js"></script>
![image](https://user-images.githubusercontent.com/41436763/216871591-f68bee06-1bba-465f-bdc3-1b0199b21e36.png)

![image](https://user-images.githubusercontent.com/41436763/216871870-7ac51f59-bda3-4ea3-bda4-6ee1a654a371.png)

**Comment by hanayik** [comment]
@wujie1314 , can you try version 0.33.0 and let us know if you are experiencing issues?

**Comment by wujie1314** [comment]
![image](https://user-images.githubusercontent.com/41436763/224866744-b60e324e-735c-4670-95f1-d0360cb89c34.png)
![image](https://user-images.githubusercontent.com/41436763/224866800-0964b409-f1cf-4da3-a067-8ad6401b12eb.png)

**Comment by neurolabusc** [comment]
@wujie1314 it is very hard to replicate your issue when you insert your code as a pixel-based png image instead of as a text-based [codeblock](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks). Regardless, I can not replicate your image. I do think you need should remove `opt:` from this line

```
var nv1 = new niivue.Niivue(opt: {isResizeCanvas: false});`
```

This may have to do with how you are sharing your images. The code below works fine if you name this `\niivue\src\index.html` and run `npm run dev` from the `\niivue` folder - npm will work as a local file server. Another way to share files is to place your images in a github repository [and share them with github pages](https://www.freecodecamp.org/news/publish-your-website-netlify-github/). An example of this method is the [niivue-demo-images](https://github.com/niivue/niivue-demo-images) which are hosted by github and therefore visible from NiiVue tools like [this demo](https://niivue.github.io/niivue/features/test_images.html).

```javascript
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>NiiVue</title>
    <link rel="icon" href="data:;base64,iVBORw0KGgo=" />
  </head>
  <style>
    html,
    body {
      background-color: white;
      color: black;
    }
    canvas {
      position: absolute;
      cursor: crosshair;
    }
  </style>
  <body>
    <main>
      <canvas id="gl1" width="640" height="640" style="width: 640px; height: 640px"></canvas>
    </main>
    <script src="https://unpkg.com/@niivue/niivue@0.33.0/dist/niivue.umd.js"></script>
    <script type="module" async>
      var volumeList1 = [
        { url: "./amos_0004_1.nii.gz", colorMap: "Gray", },
        { url: "./amos_0004_2.nii.gz", colorMap: "Red",  },
        { url: "./amos_0004_3.nii.gz", colorMap: "Green",  },
        { url: "./amos_0004_4.nii.gz", colorMap: "Blue",  },
      ];
      var nv = new niivue.Niivue({ isResizeCanvas: false });
      nv.setRadiologicalConvention(false);
      nv.attachTo("gl1");
      await nv.loadVolumes(volumeList1);
      nv.opts.isColorbar = true;
      nv.setSliceType(nv.sliceTypeRender);
    </script>
  </body>
</html>
```

![v33](https://user-images.githubusercontent.com/8930807/225156497-b0ee36d2-2013-4292-88da-3399a69128b0.png)

**Comment by wujie1314** [comment]
I copied your code and got the result
![image](https://user-images.githubusercontent.com/41436763/225188893-c8c5fa79-a576-4aa9-b5b7-fd3aee6057a9.png)

**Comment by neurolabusc** [comment]
I have created [this repository](https://github.com/neurolabusc/niivue501) that hosts this [live demo](https://neurolabusc.github.io/niivue501/). 
 - Please see if the [live demo](https://neurolabusc.github.io/niivue501/) works on your device. If it does, the problem is not with NiiVue or your browser, but with your local web server. 

You might want to do some diagnostics to identify your operating system, node version and npm version.  Here are the details for my local machine:

```bash
$ uname -a
Darwin CR.local 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:39:46 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6020 arm64
$ node --version
v18.14.2
$ npm --version
9.5.0
```

**Comment by neurolabusc** [comment]
@cdrake and @hanayik the live demo works on my MacOS laptop and iOS phone. I am traveling today but perhaps you can test on Android and Windows.

**Comment by hanayik** [comment]
> 

@neurolabusc, it does not work on my window 11 machine :(

I can debug some more and report back.

**Comment by neurolabusc** [comment]
@hanayik I bet the issue is that ANGLE's DirectX does not (yet) support the WebGL2 function copyTexSubImage3D(). If I am right, you elicit or remove this bug by comparing NiiVue before and after this commit: [Use copyTexSubImage3D() instead of pass through shader](https://github.com/niivue/niivue/commit/b854d7066f1438e832f11861d4ec21113cea2649).

**Comment by neurolabusc** [comment]
@wujie1314 this reflects a bug with ANGLE's DirectX WebGL2 function copyTexSubImage3D. I guess we can reinstate the pass through shader to restore behavior, andd we can report this to Google. In the mean time, you can fix the behavior by using OpenGL instead of DirectX  as your Chrome back end:

![niivue](https://user-images.githubusercontent.com/8930807/225458441-0a9b5259-5207-4b0a-9bac-f78b178a9a5f.png)

**Comment by neurolabusc** [comment]
Latest [commit](https://github.com/niivue/niivue/commit/4b03a82a7207dc8a8e678c695f408735905629aa) to drawing branch restores pass-through shader instead of using copyTexSubImage3D.
 - [Bug reported to ANGLE developers](https://bugs.chromium.org/p/angleproject/issues/detail?id=8085)
 - The new variable `isUseCopyTexSubImage3D` determines whether blending is done using the webGL function copyTexSubImage3D or a pass-through shader. I have set this to false. Long term, if ANGLE is patched we can set this to true. Alternatively, when we could set this when we create the WebGL texture by detecting [if the backend is DirectX](https://github.com/luruke/gl-backend/blob/main/index.js). At the moment, we use the pass through for similar behavior across all systems.

**Comment by hanayik** [comment]
@wujie1314 , please try version `0.33.1`
